### PR TITLE
[Feature][Sampling] Add native GPU output no-repeat n-gram masking

### PR DESCRIPTION
**Pull request by HuXinjing**
## Purpose

Add an output-only no-repeat n-gram GPU logits-processor plug-in for Model Runner V2. The implementation scans the existing persistent GPU token history with one batched Triton kernel, avoiding the per-request Python processor's model-runner fallback.

The processor is loaded through the existing custom logits-processor path with the FQCN `vllm.v1.worker.gpu.sample.no_repeat_ngram:NoRepeatNGramState`; it is not registered in the built-in sampler list. It is available offline through `SamplingParams.extra_args` and online through `vllm_xargs`. It accepts:

- `no_repeat_ngram_size`, with `ngram_size` as an OCR-integration compatibility alias;
- `window_size`, which bounds the recent output-token history searched;
- `whitelist_token_ids`, which keeps OCR grounding/control tokens available.

Prompt tokens are deliberately excluded. Reused request slots are reset, invalid or out-of-vocabulary values fail closed, and speculative decoding is rejected when the plug-in is initialized.

The DeepSeek-OCR, DeepSeek-OCR-2, and multi-image OCR examples now load this GPU plug-in instead of registering the legacy per-request processor. For exact compatibility, the `ngram_size=1` alias preserves the legacy processor's no-op behavior, while canonical `no_repeat_ngram_size=1` retains standard unigram blocking semantics.

## Related work / duplicate check

- #56497 provides the Model Runner V2 logits-processor lifecycle and request-state interface. This PR uses that merged custom plug-in interface and adds the GPU processor itself.
- #57618 was a DeepSeek/Unlimited-OCR custom-processor implementation stacked on #56497 and was closed without merge. This PR provides the batched Triton path and covers its requested `window_size` and `whitelist_token_ids` behavior.
- #40704 implements n-gram speculative drafting and variable-length draft plumbing. This PR implements an exact output constraint, so the execution path and semantics are different.

Duplicate searches covered `no repeat ngram`, `ngram GPU sampler`, `DeepSeek OCR ngram`, and Model Runner V2 custom logits processors.

## Validation

The complete pre-commit suite for all touched files passes, including Ruff, formatting, typos, Markdown lint, mypy, SPDX, forbidden-import, and accelerator-API checks.

Targeted GPU and sampler regression tests:

```text
tests/v1/worker/test_gpu_no_repeat_ngram.py
tests/v1/worker/test_gpu_logits_processors.py
tests/v1/worker/test_logits_processor_loader.py
tests/v1/worker/test_gpu_sampler_flags.py

43 passed
```

The CUDA tests compare the Triton kernel with a Python reference for n-gram sizes 1, 2, 3, 5, 7, 11, and 17, and directly compare it with the legacy OCR processor across repeated-token, bounded-window, and whitelist cases. They also cover output-only semantics, compacted-batch request-slot mapping, argument validation, plug-in initialization and admission, slot reuse, and the explicit legacy/canonical unigram behavior.

## Model evaluation

A controlled generative-OCR crossover ran on one H200 with the same 100 rendered pages, prompts, sampling parameters, concurrency 32, 8 warm-up requests, and three measured rounds. Both paths used the same `TRITON_ATTN` decoder backend and `TORCH_SDPA` multimodal-encoder backend on the vLLM 0.29.0 CUDA 13.0 runtime:

- official per-request Python processor: **7.9096 layout pages/s**;
- this GPU plug-in: **8.5430 layout pages/s** (**+8.0% throughput**, **-7.4% elapsed time**).

All 100 outputs from both paths retained the expected layout structure, with no empty or repeated-punctuation outputs. The plug-in run mounted this PR's Python source over the 0.29.0 runtime because no matching prebuilt PR image was available; incompatible FlashAttention Python/native ABIs were therefore avoided by holding both runs to the same backends above. This is a layout-stage comparison, not a claim of the same end-to-end document throughput gain. The repository includes the reusable HTTP benchmark driver used for the comparison.

## AI assistance

AI assistance was used for implementation, review, experiment orchestration, and drafting this description. The human submitter owns the contribution; the latest upstream rebase and Model Runner V2 adaptation require final human review before requesting full CI or merge.

### CODE REVIEW

**Comment by HuXinjing** [comment]
This PR is ready for maintainer review. Since it modifies the Model Runner V2 GPU sampling path and adds a Triton kernel, I believe it should run the full CI suite. Could a maintainer please review it for ready-run-all-tests when appropriate?

**Comment by HuXinjing** [comment]
verified

**Comment by HuXinjing** [comment]
Rebased onto the latest `upstream/main` and adapted the implementation to the Model Runner V2 logits-processor interface merged in #56497. This update also adds the requested `window_size` and `whitelist_token_ids` behavior (plus the existing `ngram_size` compatibility alias), while keeping the native batched Triton path. Local validation: all pre-commit hooks passed, and 37 targeted GPU/sampler tests passed, including processor lifecycle, loader, slot reuse, window, whitelist, and argument validation. The PR description now records the current architecture and validation results.

**Comment by HuXinjing** [comment]
Follow-up for the MRV2 OCR gap referenced from #46564/#47172: the DeepSeek-OCR, DeepSeek-OCR-2, and multi-image examples now use the built-in path directly, without registering `NGramPerReqLogitsProcessor`. I also added direct equivalence coverage against the legacy OCR processor for repeated-token, windowed, and whitelisted cases. The compatibility alias preserves the legacy `ngram_size=1` no-op, while canonical `no_repeat_ngram_size=1` keeps standard unigram semantics. Updated local validation: all touched-file pre-commit hooks passed and 42 targeted GPU/sampler tests passed.

File: vllm/v1/worker/gpu/sample/no_repeat_ngram.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            ):
+                raise ValueError("whitelist_token_ids must contain only integers")
+            if len(set(whitelist)) > MAX_WHITELIST_TOKEN_IDS:
+                raise ValueError(
+                    f"whitelist_token_ids supports at most "
+                    f"{MAX_WHITELIST_TOKEN_IDS} unique IDs"
+                )
+
+    def add_request(self, req_idx: int, sampling_params: SamplingParams) -> bool:
+        self.validate_params(sampling_params)
```
Comment (sawsa307) [inline vllm/v1/worker/gpu/sample/no_repeat_ngram.py]: We don't need to explicitly invoke validate_params here as it shall be covered by [validate_logits_processors_params()](https://github.com/vllm-project/vllm/blob/54020c3c3ec9de219929a5a956010a0251b87520/vllm/v1/engine/input_processor.py#L58).

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 2]
```diff
@@ -63,13 +64,15 @@ def __init__(
         self.penalties_state = PenaltiesState(vllm_config, lp_req_state)
         logit_bias_state = LogitBiasState(vllm_config, lp_req_state)
         bad_words_state = BadWordsState(vllm_config, lp_req_state)
+        no_repeat_ngram_state = NoRepeatNGramState(vllm_config, lp_req_state)
```
Comment (sawsa307) [inline vllm/v1/worker/gpu/sample/sampler.py]: Currently we consider this feature a plug-in logits processor, so it should be wired through the custom path — FQCN via `--logits-processors vllm.v1.worker.gpu.sample.no_repeat_ngram:NoRepeatNGramState`, or an entrypoint instead of being added to self.logits_processors directly. The built-in list should stay minimal.

File: vllm/v1/worker/gpu/sample/no_repeat_ngram.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        if whitelist and (
+            whitelist[0] < 0 or whitelist[-1] >= self.req_states.vocab_size
+        ):
+            raise ValueError("whitelist_token_ids contains an out-of-vocabulary ID")
+        self.ngram_sizes.np[req_idx] = raw_size
+        self.window_sizes.np[req_idx] = window_size
+        self.whitelist_lens.np[req_idx] = len(whitelist)
+        if whitelist:
+            self.whitelist_token_ids.stage_write(req_idx, 0, whitelist)
+        self._ever_enabled = True
```
Comment (sawsa307) [inline vllm/v1/worker/gpu/sample/no_repeat_ngram.py]: Can you explain a bit more on what's `self._ever_enabled` for? It is initialized as False but it is a one-way flag. What's the additional optimization it gives on top of the np check?

File: vllm/v1/worker/gpu/sample/no_repeat_ngram.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        # compatibility alias while canonical no_repeat_ngram_size=1 keeps the
+        # standard unigram constraint.
+        if native_size is None and compatible_size == 1:
+            self.ngram_sizes.np[req_idx] = 0
+            self.window_sizes.np[req_idx] = 0
+            self.whitelist_lens.np[req_idx] = 0
+            return False
+        if raw_size > self.max_model_len:
+            raise ValueError("ngram size cannot exceed the configured max model length")
+        if self.speculative:
```
Comment (sawsa307) [inline vllm/v1/worker/gpu/sample/no_repeat_ngram.py]: Can we move this check into __init__()? Basically we read its value in constructor but I don't see any point delaying the error until during add_request()

File: docs/features/custom_arguments.md [hunk 5]
```diff
@@ -47,3 +47,42 @@ batch = await client.completions.create(
 
 !!! note
     `vllm_xargs` is assigned to `SamplingParams.extra_args` under the hood, so code which uses `SamplingParams.extra_args` is compatible with both offline and online scenarios.
+
```
Comment (sawsa307) [inline docs/features/custom_arguments.md:50]: Please update the doc accordingly as we want n-gram to be a custom plug-in.

**Review by sawsa307 — COMMENTED** [review]
Thank you @HuXinjing for the contribution!
I took a look and leave some comments. Overall LGTM. Once they are addressed I'll leave it to @njhill to take a final pass.

File: vllm/v1/worker/gpu/sample/no_repeat_ngram.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    Token history already lives in :class:`RequestState`; retaining a second
+    Python or GPU index costs more than scanning it in parallel for the long
+    n-grams used by generative OCR models. The only additional persistent
+    state is therefore the constraint configuration for each request slot.
+    """
+
+    def __init__(self, vllm_config: "VllmConfig", req_states: LogitsProcRequestState):
+        self.req_states = req_states
+        self.max_model_len = req_states.all_token_ids.gpu.shape[1]
+        self.speculative = getattr(vllm_config, "speculative_config", None) is not None
```
Comment (sawsa307) [inline vllm/v1/worker/gpu/sample/no_repeat_ngram.py]: We can do ``vllm_config.speculative_config is not None`?

**Comment by HuXinjing** [comment]
@sawsa307 Thank you for the thoughtful review. I have updated the PR to address your comments:

- the processor is now loaded only through the custom FQCN path and is no longer added to the built-in sampler list;
- the redundant `validate_params()` call in `add_request()` was removed;
- `_ever_enabled` was removed, leaving the request-slot NumPy check as the single fast-path guard;
- the speculative-decoding check now runs in `__init__()` and directly reads `vllm_config.speculative_config`;
- the documentation, OCR examples, and benchmark instructions now consistently describe and use the plug-in path.

I reran the affected pre-commit checks and targeted tests: **43 passed**. I also reran the 100-page OCR comparison on an H200 with both paths using the same `TRITON_ATTN` and `TORCH_SDPA` backends. The official per-request processor averaged 7.9096 pages/s, while the GPU plug-in averaged 8.5430 pages/s (+8.0%); all 100 outputs from both paths retained valid layout structure.

The branch and PR description have been updated. When you have a chance, could you please take another look and let me know whether the changes now match your expectations?

File: tests/v1/worker/test_gpu_sampler_flags.py [hunk 7]
```diff
@@ -17,6 +17,7 @@
     LogitsContext,
     LogitsProcessor,
 )
```
Comment (sawsa307) [inline tests/v1/worker/test_gpu_sampler_flags.py:20]:
```suggestion
    LogitsContext,
    LogitsProcessor,
    LogitsProcRequestState,
)
```

File: tests/v1/worker/test_gpu_sampler_flags.py [hunk 8]
```diff
@@ -39,8 +42,13 @@ def _make_sampler(custom_logits_processors: Sequence[LogitsProcessor] = ()) -> S
         vocab_size=VOCAB_SIZE,
         device=DEVICE,
     )
+    vllm_config = SimpleNamespace(
+        reasoning_config=MockReasoningConfig(), speculative_config=None
+    )
+    if custom_logits_processors is None:
+        custom_logits_processors = [NoRepeatNGramState(vllm_config, req_states)]
```
Comment (sawsa307) [inline tests/v1/worker/test_gpu_sampler_flags.py]:
```suggestion
        lp_req_state = LogitsProcRequestState.from_request_state(req_states)
        custom_logits_processors = [NoRepeatNGramState(vllm_config, lp_req_state)]
```

File: tests/v1/worker/test_gpu_sampler_flags.py [hunk 9]
```diff
@@ -86,6 +109,29 @@ def test_logits_processing_cache_is_overwritten_when_slot_is_reused():
...
+        max_model_len=64,
+        max_num_batched_tokens=16,
+        num_speculative_steps=1,
+        vocab_size=VOCAB_SIZE,
+        device=DEVICE,
+    )
+    vllm_config = SimpleNamespace(speculative_config=object())
+
+    with pytest.raises(ValueError, match="does not support speculative decoding"):
+        NoRepeatNGramState(vllm_config, req_states)
```
Comment (sawsa307) [inline tests/v1/worker/test_gpu_sampler_flags.py]:
```suggestion
        lp_req_state = LogitsProcRequestState.from_request_state(req_states)
        NoRepeatNGramState(vllm_config, lp_req_state)
```

**Review by sawsa307 — COMMENTED** [review]
One last nit on test wiring: custom logits processors receive the narrow `LogitsProcRequestState` at construction, not the internal `RequestState`. Going through `LogitsProcRequestState.from_request_state()` keeps the tests on the runtime contract, matching the pattern in `test_gpu_logit_bias.py` / `test_gpu_bad_words.py`. Inline suggestions below.
@njhill this PR should be ready for a final pass review, thanks!

File: tests/v1/worker/test_gpu_sampler_flags.py (hunk 7 above)
Comment (HuXinjing) [reply tests/v1/worker/test_gpu_sampler_flags.py:20]: Applied, thank you. The tests now import `LogitsProcRequestState` and construct the plug-in through `LogitsProcRequestState.from_request_state()`, matching the runtime contract. The targeted suite still passes: 43 tests.

**Review by sawsa307 — COMMENTED** [review]
Thanks for working on this! I had a discussion with @njhill and since DeepSeek-OCR is currently the only consumer of this feature, we'd prefer to keep the processor model-scoped instead of adding a generic plugin under `vllm/v1/worker/gpu/sample/`. And as Model Runner V2 supports custom logits processors (#56497), the new implementation should serve as an enhancement implements it with the V2 interface rather than the legacy V1 per-request path. For testing we would need to make sure:

1. This custom LP is supported with MRV2
2. Existing perf comparison

Concretely:

1. **Move the implementation into the model file.** Relocate the V2 logits processor class (including the Triton kernel) from `vllm/v1/worker/gpu/sample/no_repeat_ngram.py` into `vllm/model_executor/models/deepseek_ocr.py`, next to the existing V1 `NGramPerReqLogitsProcessor` (`deepseek_ocr.py:140`). Keep the `triton` import function-scoped (inside `__init__`/`apply`): the frontend process imports this module to run `validate_params` at request admission (`vllm/v1/worker/gpu/sample/logits_processor/loader.py`), so the module must stay importable without model-runner/triton dependencies.

3. **Auto-register it via a model config hook, following the Unlimited-OCR pattern (#46564).** Add a `DeepseekOCRForCausalLMConfig(VerifyAndUpdateConfig)` (and the DeepSeek-OCR-2 arch) to `MODELS_CONFIG_MAP` in `vllm/model_executor/models/config.py`, mirroring `UnlimitedOCRForCausalLMConfig`. In `verify_and_update_config(vllm_config)`, append the processor to `model_config.logits_processors`, selecting the class by runner version: the V2 GPU class when `vllm_config.use_v2_model_runner` is true, otherwise keep registering the existing V1 `NGramPerReqLogitsProcessor` (the V1 and V2 loaders each reject classes of the other interface, so both runner paths need a matching class). Also dedupe against user-supplied entries in `model_config.logits_processors` so explicit registration doesn't apply the processor twice.

4. **Keep activation per-request.** With the class auto-registered, users no longer pass `logits_processors=[...]` in `EngineArgs`; activation stays via `extra_args` (`ngram_size` / `no_repeat_ngram_size`, `window_size`, `whitelist_token_ids`). When `ngram_size` is absent, `add_request()` should return `False` so default sampling behavior is unchanged.

5. **Keep speculative decoding unsupported.** The feature functionally does not support spec decode (the V1 path already rejects custom logits processors with spec decode via `STR_SPEC_DEC_REJECTS_LOGITSPROCS`, and this PR does the same at plugin init), so keep rejecting it outright when the processor is registered. Note the corollary of auto-registration: DeepSeek-OCR + spec decode becomes an unsupported combination entirely, which matches the feature's semantics.

6. **Update examples and docs.** Drop the explicit `logits_processors=[NGramPerReqLogitsProcessor]` from the DeepSeek-OCR / multi-image OCR examples (`examples/generate/multimodal/vision_language_offline.py`, `vision_language_multi_image_offline.py`); keep the `extra_args`. Adjust `docs/features/custom_arguments.md` accordingly.

7. **Tests.** Update the import path in `tests/v1/worker/test_gpu_no_repeat_ngram.py`, and add coverage for the auto-registration: V1 runner registers the V1 class, V2 runner registers the V2 class, and duplicate registration is a no-op.
With this shape the feature stays entirely inside DeepSeek-OCR's model code and config hook, works out of the box (zero engine-level config), and doesn't extend the generic V2 sampler surface for a single-model use case.
